src/Lean/Elab/Command.lean

```diff
@@ -12,17 +12,35 @@ import Lean.Language.Basic
 
 namespace Lean.Elab.Command
 
+/--
+A `Scope` collects context information which is inherently scoped: this contains
+the current namespace and information about `open`, `universe` or `variable` declarations.
+There is always a base scope; scopes can be nested.
+
+`universe`, `open` and `variable` declarations only modify the current scope
+(and all scopes nested below it): these declarations usually *append* to the
+context in the current scope; `variable` statements can also change existing items.
+(XXX Is shadowing possible? If so, this sentence should be reworded.)
+Unclosed scopes are automatically closed at the end of the current module.
+-/
 structure Scope where
   header        : String
   opts          : Options := {}
+  /-- The current namespace: by default, this is an anonymous namespace -/
   currNamespace : Name := Name.anonymous
+  /-- All currently `open`ed namespaces -/
   openDecls     : List OpenDecl := []
+  /-- All universe levels which apply in the current scope -/
   levelNames    : List Name := []
-  /-- section variables -/
+  /-- All variable binders which apply in the current scope -/
   varDecls      : Array (TSyntax ``Parser.Term.bracketedBinder) := #[]
   /-- Globally unique internal identifiers for the `varDecls` -/
   varUIds       : Array Name := #[]
-  /-- noncomputable sections automatically add the `noncomputable` modifier to any declaration we cannot generate code for. -/
+  /-- Whether the current scope is (or is nested inside) a `noncomputable` `section`:
+  in this case, automatically add the `noncomputable` modifier to any declaration
+  we cannot generate code for.
+  A scope is noncomputable if it is a noncomputable section or its parent is noncomputable.
```

Comment:
The first and second sentence are somewhat redundant, as I liked both phrasings and could not decide if any one was better. Feel free to articulate a preference (or keep both).

Reply:
Could you make a version that doesn't talk about nesting scopes? For someone using this type, it's important that they know that none of the other scopes are looked at when deciding whether it's noncomputable, so I'd avoid talking about parent/child scopes too much, to prevent it giving the wrong idea.

Here's a suggestion to give some ideas:
```quote
Whether declarations that fail to compile should automatically receive the `noncomputable` modifier.
A scope with this flag set is created by `noncomputable section`.

Recall that a new scope inherits all values from its parent scope, so all sections and namespaces nested within a `noncomputable` section will also have this flag set.
```

Reply:
Makes sense - I like your suggestion, and went with something very similar.